### PR TITLE
Feature/boundary conditions

### PR DESCRIPTION
**Pull request by stonecoldhughes**
Please review the [developer documentation](https://github.com/project-asgard/asgard/wiki/developing)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [ ] Bugfix
- [ ] New feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [ ] Yes
- [ ] No

## Checklist

_Put an x in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- [ ] this PR is up to date with current the current state of 'develop'
- [ ] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [ ] documentation has been added (if appropriate)

### CODE REVIEW

**Comment by bmcdanie** [comment]
test this please

File: CMakeLists.txt [hunk 1]
```diff
@@ -19,7 +19,7 @@ set (CMAKE_CXX_STANDARD_REQUIRED ON)
 message (STATUS "CMAKE_BUILD_TYPE is ${CMAKE_BUILD_TYPE}")
 
 # add compiler flags we always want to use
-add_compile_options ("-Wall" "-Wextra" "-Wpedantic" "-g" "-O3")
+add_compile_options ("-Wall" "-Wextra" "-Wpedantic" "-O3")
```
Comment (bmcdanie) [inline CMakeLists.txt]: give my flag back

File: src/boundary_conditions.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
+#pragma once
+#include "element_table.hpp"
+#include "fast_math.hpp"
+#include "matlab_utilities.hpp"
+#include "pde/pde_base.hpp"
+#include "transformations.hpp"
+
+template<typename P>
+class bc_timestepper
```
Comment (bmcdanie) [inline src/boundary_conditions.hpp]: no state between functions - can we restructure to free functions in a component namespace?

File: src/boundary_conditions.cpp [hunk 3]
```diff
@@ -0,0 +1,278 @@
+#include "boundary_conditions.hpp"
+#include <cstdio>
+
+template<typename P>
+bc_timestepper<P>::bc_timestepper(PDE<P> const &pde, element_table const &table,
```
Comment (bmcdanie) [inline src/boundary_conditions.cpp]: please leave brief comment on the two major functions re: input/outputs and general purpose

File: src/boundary_conditions.cpp [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+        if (p_term.right_homo == homogeneity::inhomogeneous)
+        {
+          fk::vector<P> trace_bc = compute_right_boundary_condition(
+              p_term.g_func, t_init, d.get_level(), d.get_degree(),
+              d.domain_min, d.domain_max, p_term.right_bc_funcs[dim_num]);
+
+          std::vector<fk::vector<P>> p_term_right_bcs = generate_partial_bcs(
+              dimensions, dim_num, p_term.right_bc_funcs, t_init, partial_terms,
+              p_num, std::move(trace_bc));
```
Comment (bmcdanie) [inline src/boundary_conditions.cpp:81]: why move here?

File: src/boundary_conditions.cpp [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+bc_timestepper<P>::bc_timestepper(PDE<P> const &pde, element_table const &table,
+                                  int const start_element,
+                                  int const stop_element, P const t_init)
+    : t_init(t_init),
+      bc_size((stop_element - start_element + 1) *
+              std::pow(pde.get_dimensions()[0].get_degree(), pde.num_dims)),
+      pde(pde)
+{
+  assert(bc_size > 0);
+
```
Comment (bmcdanie) [inline src/boundary_conditions.cpp:26]: lets add some asserts, e.g. on stop/start element

File: src/boundary_conditions_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+}
+
+template<typename P>
+void test_compute_boundary_condition(PDE<P> &pde,
+                                     std::string gold_filename_prefix,
+                                     double const eps_multiplier)
+{
+  generate_all_coefficients<P>(pde);
+
+  /* Open the file and feed params to compute_boundary_conditions */
```
Comment (bmcdanie) [inline src/boundary_conditions_tests.cpp]: open what file?

File: src/boundary_conditions_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,318 @@
+#include "boundary_conditions.hpp"
+#include "coefficients.hpp"
+#include "element_table.hpp"
+#include "tests_general.hpp"
+
+template<typename P>
+void test_boundary_condition_vector(PDE<P> &pde,
+                                    std::string gold_filename_prefix,
```
Comment (bmcdanie) [inline src/boundary_conditions_tests.cpp]: std::string const

File: src/boundary_conditions_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+          relaxed_comparison(gold_right_bc_vector, right_bc, eps_multiplier);
+        }
+      }
+    }
+  }
+
+  return;
+}
+
+TEMPLATE_TEST_CASE("problem separability", "", double, float)
```
Comment (bmcdanie) [inline src/boundary_conditions_tests.cpp]: for all these, the second string here should be component name in brackets `[boundary_condition]`

File: src/boundary_conditions_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+    int const level  = 2;
+    int const degree = 2;
+    auto const pde   = make_PDE<TestType>(PDE_opts::diffusion_1, level, degree);
+
+    std::string const gold_filename_prefix = "../testing/generated-inputs/"
+                                             "compute_boundary_conditions/"
+                                             "diffusion1/";
+
+    double epsilons = 0;
+    if constexpr (std::is_same<TestType, float>::value == true)
```
Comment (bmcdanie) [inline src/boundary_conditions_tests.cpp]: don't do `boolean==true`, just `if constexpr(std::is_same<TestType,float>::value)`

also, let's call these `eps_scale` or `tol_factor` - not epsilon. epsilon means something else.

Comment (bmcdanie) [reply src/boundary_conditions_tests.cpp]: also, pull them out to the `TEMPLATE_TEST_CASE` level if they're consistent across sections.

small nitpick, just trying to keep tests concise

File: src/pde/pde_base.hpp [hunk 10]
```diff
@@ -141,24 +147,40 @@ enum class flux_type
 // do dimensions own terms? need dimension info in
 // term construction...
 
+/* encapsulates all boundary condition information for a partial term - decide
+   whether to use later */
```
Comment (bmcdanie) [inline src/pde/pde_base.hpp]: what do you mean here?

File: src/time_advance.cpp [hunk 11]
```diff
@@ -47,8 +49,15 @@ void explicit_time_advance(PDE<P> const &pde, element_table const &table,
 
   apply_A(pde, table, grid, chunks, host_space, dev_space, batches);
   reduce_results(host_space.fx, host_space.reduced_fx, plan, my_rank);
+  /* the call below calculates "source1" from matlab - line 32 time_advance.m */
```
Comment (bmcdanie) [inline src/time_advance.cpp]: good to tie these together, but this comment is unlikely to age well - when referencing the other codebase, drop a permalink

File: src/time_advance.cpp [hunk 12]
```diff
@@ -47,8 +49,15 @@ void explicit_time_advance(PDE<P> const &pde, element_table const &table,
...
   scale_sources(pde, unscaled_sources, host_space.scaled_source, time);
   fm::axpy(host_space.scaled_source, host_space.reduced_fx);
+
+  fk::vector<P> bc0 = bc_generator.advance(time);
+  fm::axpy(bc0, host_space.reduced_fx);
+  /* end new code */
```
Comment (bmcdanie) [inline src/time_advance.cpp]: remove

File: src/time_advance.cpp [hunk 13]
```diff
@@ -60,13 +69,19 @@ void explicit_time_advance(PDE<P> const &pde, element_table const &table,
...
+
   exchange_results(host_space.reduced_fx, host_space.result_2, elem_size, plan,
                    my_rank);
 
   fm::copy(host_space.x_orig, host_space.x);
   P const fx_scale_2a = a31 * dt;
   P const fx_scale_2b = a32 * dt;
 
+  /* result_1 is k1, result_2 is k2 */
```
Comment (bmcdanie) [inline src/time_advance.cpp]: flesh out these comments or remove them

File: src/time_advance.cpp [hunk 14]
```diff
@@ -47,8 +49,15 @@ void explicit_time_advance(PDE<P> const &pde, element_table const &table,
 
   apply_A(pde, table, grid, chunks, host_space, dev_space, batches);
   reduce_results(host_space.fx, host_space.reduced_fx, plan, my_rank);
+  /* the call below calculates "source1" from matlab - line 32 time_advance.m */
   scale_sources(pde, unscaled_sources, host_space.scaled_source, time);
   fm::axpy(host_space.scaled_source, host_space.reduced_fx);
+
+  fk::vector<P> bc0 = bc_generator.advance(time);
```
Comment (bmcdanie) [inline src/time_advance.cpp]: const?

File: src/pde/pde_base.hpp [hunk 15]
```diff
@@ -173,6 +195,13 @@ class partial_term
 
   boundary_condition const right;
 
+  homogeneity const left_homo;
```
Comment (bmcdanie) [inline src/pde/pde_base.hpp:196]: are there conditions on these, e.g. should we always have bc functions if we have inhomog boundary conditions? if we can check these for correctness with asserts in pde base class constructor, we should

File: src/pde.hpp [hunk 16]
```diff
@@ -69,6 +73,7 @@ static pde_map_t const pde_mapping = {
     {"fokkerplanck_1d_4p4", PDE_opts::fokkerplanck_1d_4p4},
     {"fokkerplanck_1d_4p5", PDE_opts::fokkerplanck_1d_4p5},
     {"fokkerplanck_2d_complete", PDE_opts::fokkerplanck_2d_complete},
+    {"diffusion_1", PDE_opts::diffusion_1},
```
Comment (bmcdanie) [inline src/pde.hpp:76]: no mapping for diffusion_2?

File: src/coefficients_tests.cpp [hunk 17]
```diff
@@ -48,7 +162,7 @@ TEMPLATE_TEST_CASE("continuity 1 (single term)", "[coefficients]", double,
       continuity1->get_coefficients(0, 0).clone_onto_host();
 
   // determined empirically 11/19
-  auto const eps_multiplier = 1e2;
+  TestType const eps_multiplier = 1e2;
```
Comment (bmcdanie) [inline src/coefficients_tests.cpp]: why?

File: src/coefficients_tests.cpp [hunk 18]
```diff
@@ -33,6 +33,120 @@ void test_coefficients(PDE<P> &pde, std::string const gold_path,
...
+  {
+    int const level  = 2;
+    int const degree = 2;
+
+    auto pde = make_PDE<TestType>(PDE_opts::diffusion_1, level, degree);
+    std::string const gold_path =
+        "../testing/generated-inputs/coefficients/diffusion1/coefficients";
+
+    double epsilons = 0;
+    if constexpr (std::is_same<TestType, float>::value == true)
```
Comment (bmcdanie) [inline src/coefficients_tests.cpp]: see other comment about not calling these epsilon.

also, can we compress? trying to keep test code concise.

`TestType epsilon = std::is_same<TestType, float>::value ? 1e3 : 1e3;`

File: src/time_advance_tests.cpp [hunk 19]
```diff
@@ -29,7 +29,7 @@ void time_advance_test(int const level, int const degree, PDE<P> &pde,
                        int const num_steps, std::string const filepath,
                        bool const full_grid                            = false,
                        std::vector<std::string> const &additional_args = {},
-                       double const eps_multiplier                     = 1e4)
+                       double const eps_multiplier                     = 3e6)
```
Comment (bmcdanie) [inline src/time_advance_tests.cpp]: this should probably be changed back - you adjusted the default tolerance other tests are using, but it looks like you're passing a tolerance for your new pdes?

File: src/transformations.cpp [hunk 20]
```diff
@@ -233,8 +233,10 @@ combine_dimensions(int const degree, element_table const &table,
     int const stop_index  = start_index + std::pow(degree, num_dims) - 1;
     fk::vector<P, mem_type::view> combined_view(combined, start_index,
                                                 stop_index);
+
```
Comment (bmcdanie) [inline src/transformations.cpp:236]: well ok I suppose

File: src/boundary_conditions.cpp [hunk 21]
```diff
@@ -0,0 +1,278 @@
...
+  {
+    partial_bc_vecs.emplace_back(
+        forward_transform(dimensions[dim_num], bc_funcs[dim_num], time));
+  }
+
+  return partial_bc_vecs;
+}
+
+template<typename P>
+fk::vector<P> bc_timestepper<P>::compute_right_boundary_condition(
```
Comment (bmcdanie) [inline src/boundary_conditions.cpp]: can we merge this function with the one beneath? 

at quick glance, they only differ in one or two lines.

File: src/boundary_conditions.hpp [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+{
+public:
+  bc_timestepper(PDE<P> const &pde, element_table const &table,
+                 int const start_element, int const stop_element,
+                 P const t_init = 0);
+
+  fk::vector<P> advance(P const time) const;
+
+  void print_left() const;
```
Comment (bmcdanie) [inline src/boundary_conditions.hpp]: do we still need these?

**Review by bmcdanie — CHANGES_REQUESTED** [review]
good work.

major point - let's not have another class here that must be passed into time advance. these functions require no shared state (beyond the vectors) and we can place those in the `host_workspace`.

so, please refactor to two free functions in a namespace - something like `generate` and `scale`/`scale_at_time`. advance is a little vague.

core of functions look pretty good, made some syntax nitpicks/suggestions here and there, but the above is the main point.

File: CMakeLists.txt (hunk 1 above)
Comment (stonecoldhughes) [reply CMakeLists.txt]: *red team has acquired the flag*

File: src/boundary_conditions.hpp (hunk 2 above)
Comment (stonecoldhughes) [reply src/boundary_conditions.hpp]: aye aye

File: src/boundary_conditions.cpp (hunk 3 above)
Comment (stonecoldhughes) [reply src/boundary_conditions.cpp]: commented

File: src/boundary_conditions.cpp (hunk 4 above)
Comment (stonecoldhughes) [reply src/boundary_conditions.cpp:81]: That function returns a container of `fk::vector` and `trace_bc` will be somewhere in that container - where depends on loop iteration variables.

File: src/boundary_conditions.cpp (hunk 5 above)
Comment (stonecoldhughes) [reply src/boundary_conditions.cpp:26]: done

File: src/boundary_conditions_tests.cpp (hunk 6 above)
Comment (stonecoldhughes) [reply src/boundary_conditions_tests.cpp]: extraneous comment removed

File: src/boundary_conditions_tests.cpp (hunk 7 above)
Comment (stonecoldhughes) [reply src/boundary_conditions_tests.cpp]: std::string const &

File: src/boundary_conditions_tests.cpp (hunk 8 above)
Comment (stonecoldhughes) [reply src/boundary_conditions_tests.cpp]: done

File: src/boundary_conditions_tests.cpp (hunk 7 above)
Comment (bmcdanie) [reply src/boundary_conditions_tests.cpp]: even better

File: src/boundary_conditions_tests.cpp (hunk 9 above)
Comment (stonecoldhughes) [reply src/boundary_conditions_tests.cpp]: done

File: src/boundary_conditions.cpp [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+  assert(destination_slice.size() == legendre_polys_at_value.size());
+
+  destination_slice = fk::vector<P>(legendre_polys_at_value);
+
+  return bc;
+}
+
+template<typename P>
+fk::vector<P> bc_timestepper<P>::compute_left_boundary_condition(
+    g_func_type const g_func, P const time, int level, int degree,
```
Comment (bmcdanie) [inline src/boundary_conditions.cpp]: int const level, int const degree

File: src/boundary_conditions.cpp [hunk 24]
```diff
@@ -0,0 +1,278 @@
...
+              combine_dimensions(d.get_degree(), table, start_element,
+                                 stop_element, p_term_left_bcs);
+
+          left_pvecs.emplace_back(std::move(combined));
+        }
+
+        if (p_term.right_homo == homogeneity::inhomogeneous)
+        {
+          fk::vector<P> trace_bc = compute_right_boundary_condition(
+              p_term.g_func, t_init, d.get_level(), d.get_degree(),
```
Comment (bmcdanie) [inline src/boundary_conditions.cpp]: passing four `dimension` fields here - just pass the dimension as a ref

File: src/boundary_conditions.cpp [hunk 25]
```diff
@@ -0,0 +1,278 @@
...
+  P const scale_factor = (1.0 / std::sqrt(domain_per_cell)) *
+                         bc_func(fk::vector<P>({domain_max}), time)(0) * g;
+
+  legendre_polys_at_value.scale(scale_factor);
+
+  int const start_index = degree * (total_cells - 1);
+  int const stop_index  = degree * total_cells - 1;
+  fk::vector<P, mem_type::view> destination_slice(bc, start_index, stop_index);
+
+  assert(destination_slice.size() == legendre_polys_at_value.size());
```
Comment (bmcdanie) [inline src/boundary_conditions.cpp]: this assert is extraneous - the copy below won't work if the sizes don't match

File: src/pde/pde_base.hpp (hunk 10 above)
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp]: extraneous comment removed

File: src/time_advance.cpp (hunk 11 above)
Comment (stonecoldhughes) [reply src/time_advance.cpp]: done

File: src/time_advance.cpp (hunk 12 above)
Comment (stonecoldhughes) [reply src/time_advance.cpp]: done

File: src/time_advance.cpp (hunk 13 above)
Comment (stonecoldhughes) [reply src/time_advance.cpp]: done

File: src/time_advance.cpp (hunk 14 above)
Comment (stonecoldhughes) [reply src/time_advance.cpp]: done

File: src/boundary_conditions.cpp (hunk 25 above)
Comment (stonecoldhughes) [reply src/boundary_conditions.cpp]: removed

File: src/boundary_conditions.cpp (hunk 24 above)
Comment (stonecoldhughes) [reply src/boundary_conditions.cpp]: done

File: src/boundary_conditions.cpp (hunk 23 above)
Comment (stonecoldhughes) [reply src/boundary_conditions.cpp]: done

File: src/boundary_conditions.hpp (hunk 22 above)
Comment (stonecoldhughes) [reply src/boundary_conditions.hpp]: removed

File: src/boundary_conditions.cpp (hunk 21 above)
Comment (stonecoldhughes) [reply src/boundary_conditions.cpp]: Not really - there are several small differences. To parameterize them, I would have to pull so much out of the function that it wouldn't even make sense for it to be a standalone function anymore

File: src/time_advance_tests.cpp (hunk 19 above)
Comment (stonecoldhughes) [reply src/time_advance_tests.cpp]: reverted

File: src/coefficients_tests.cpp (hunk 17 above)
Comment (stonecoldhughes) [reply src/coefficients_tests.cpp]: these should all be `double` because they are passed down to `relaxed_comparison` which expects a double

Comment (stonecoldhughes) [reply src/coefficients_tests.cpp]: fixed

File: src/pde.hpp (hunk 16 above)
Comment (stonecoldhughes) [reply src/pde.hpp:76]: added

File: src/coefficients_tests.cpp (hunk 18 above)
Comment (stonecoldhughes) [reply src/coefficients_tests.cpp]: done

File: src/pde/pde_base.hpp (hunk 15 above)
Comment (stonecoldhughes) [reply src/pde/pde_base.hpp:196]: done

File: src/boundary_conditions.cpp (hunk 5 above)
Comment (bmcdanie) [reply src/boundary_conditions.cpp:26]: can we have 

`assert(start_element > 0)`, `assert(stop_element < table.size()`, `assert(stop_element >= start_element)` ?

File: CMakeLists.txt [hunk 26]
```diff
@@ -19,7 +19,7 @@ set (CMAKE_CXX_STANDARD_REQUIRED ON)
 message (STATUS "CMAKE_BUILD_TYPE is ${CMAKE_BUILD_TYPE}")
 
 # add compiler flags we always want to use
-add_compile_options ("-Wall" "-Wextra" "-Wpedantic" "-g" "-O3")
+add_compile_options ("-Wall" "-Wextra" "-Wpedantic" "-O3" "-g")
```
Comment (bmcdanie) [inline CMakeLists.txt:22]: ok I guess

File: src/boundary_conditions.cpp [hunk 27]
```diff
@@ -0,0 +1,355 @@
...
+
+the fk::vectors contained in these are generated from the partial terms in the
+1D terms that are inside multi-dimensional terms in the PDE.
+
+These two outputs need only be calculated once and can then be used at any time
+value "t" to generate the complete boundary condition vector at time "t".
+
+*/
+template<typename P>
+void boundary_conditions::init_bc_parts(
```
Comment (bmcdanie) [inline src/boundary_conditions.cpp]: for these names:

init_bc_parts ->make_unscaled_bc , generate_bc -> get_scaled_bc . or something like this. we're making these at one point and scaling them for time in another. don't have to match my words exactly but try to name them in ways that show how they interact.

File: src/boundary_conditions.cpp [hunk 28]
```diff
@@ -0,0 +1,355 @@
...
+    PDE<P> const &pde, element_table const &table, int const start_element,
+    int const stop_element,
+    std::vector<std::vector<std::vector<fk::vector<P>>>> &left_bc_parts,
```
Comment (bmcdanie) [inline src/boundary_conditions.cpp]: final note: instead of in/out parameters, can we create these in the function and return them by value?

first, name them something like `using unscaled_bc_parts = std::vector..std::vector...std::vector..fk::vector;`

then use an array literal/structured binding like: https://github.com/project-asgard/asgard/blob/7dca6e78fbcbb66aebebc1b940f7245fef2de36b/src/quadrature.cpp#L134. to return them both.

Comment (bmcdanie) [reply src/boundary_conditions.cpp]: I'm going to change these in the batch refactor to map pterms directly to fk::vectors because this nested structure is confusing, but just put in the using for now.

**Review by bmcdanie — CHANGES_REQUESTED** [review]

File: src/boundary_conditions.cpp (hunk 5 above)
Comment (stonecoldhughes) [reply src/boundary_conditions.cpp:26]: done

File: src/boundary_conditions.cpp (hunk 27 above)
Comment (stonecoldhughes) [reply src/boundary_conditions.cpp]: done

File: src/boundary_conditions.cpp (hunk 28 above)
Comment (stonecoldhughes) [reply src/boundary_conditions.cpp]: done

**Review by bmcdanie — APPROVED** [review]
